design/DESIGN_typed_context_pipeline.md

```diff
@@ -0,0 +1,451 @@
+    final value = _expando[requestContext.token];
+    if (value == null) {
+      throw StateError(
+          'ContextProperty value not found. Property: ${_debugName ?? T.toString()}. '
+          'Ensure middleware has set this value for the request token.');
+    }
+    return value;
+  }
+
+  T? getOrNull(RequestContext requestContext) {
+    return _expando[requestContext.token];
+  }
+
+  void set(RequestContext requestContext, T value) {
+    _expando[requestContext.token] = value;
+  }
+
+  bool exists(RequestContext requestContext) {
+    return _expando[requestContext.token] != null;
+  }
+
+  void clear(RequestContext requestContext) {
+    _expando[requestContext.token] = null; // Clears the association in Expando
+  }
+}
+```
+Modules responsible for specific pieces of context data (e.g., `User`, `Session`) will define a private static `ContextProperty<DataType>` instance.
+
+### 2.4. Extension Types for Context Views
+
+Extension types are zero-cost abstractions (wrappers) over an instance of Relic's `RequestContext`. They provide a type-safe "view" or "contract" for accessing and attaching specific data, using `ContextProperty` instances internally.
+
+Each view typically provides:
+*   Getters for the data it represents (e.g., `UserContextView.user`).
+*   Methods to attach or set its data (e.g., `UserContextView.attachUser(User user)`). These methods use the appropriate `ContextProperty` and the `requestToken`.
+
+```dart
+// Define a ContextProperty for User data.
+// This would typically be a static final field, private to its library,
+// in a relevant class or top-level.
+final _userProperty = ContextProperty<User>('relic.auth.user');
+
+// Base view that all request contexts can be seen as initially.
+// It wraps Relic's RequestContext.
+extension type BaseContextView(RequestContext _relicContext) {
+  Request get request => _relicContext.request;
+}
+
+// A view indicating that User information is available.
+extension type UserContextView(RequestContext _relicContext) implements BaseContextView {
+  User get user => _userProperty.get(_relicContext);
+
+  void attachUser(User user) {
+    _userProperty.set(_relicContext, user);
+  }
+
+  // Optional: to check for user presence or get a nullable user
+  User? get userOrNull => _userProperty.getOrNull(_relicContext);
+  bool get hasUser => _userProperty.exists(_relicContext);
+}
+
+// A view indicating that Session information is available.
+extension type SessionContextView(RequestContext _relicContext) implements BaseContextView {
+  Session get session => _sessionProperty.get(_relicContext);
+
+  void attachSession(Session session) {
+    _sessionProperty.set(_relicContext, session);
+  }
+}
+final _sessionProperty = ContextProperty<Session>('relic.session');
+
+// A composite view indicating both User and Session information are available.
+extension type UserSessionContextView(RequestContext _relicContext) implements UserContextView, SessionContextView {
```

Comment:
I think this is the Achilles' heel of the proposal.

While the application could easily build "the one full context" (per router), how would this combine with externally provided middle?

Wouldn't I have to wrap each step in a `*ContextView` that threads through all the previous one (specific to my current app configuration)?